### DISCUSSION TITLE
How to migrate from the POJO runtime to the OSGI runtime?

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#792] [Q&A] [answered]
The POJO runtime isn't working at the moment. As described in #781, a fix will be available not before 03/2022.

So I tried how easy it is to migrate an existing POJO runtime application to an OSGI runtime application.

The basic principle is to use unzip the OSGI runtime instead of the POJO runtime (of course) and to add a call to `engineConfig.setBIRTHome(birt_home);` before calling `Platform.startup(engineConfig);` - where `birt_home` is the directory ReportEngine/platfrom (which did not exist in the POJO runtime).
This is the only thing that needs to be changes in the program source code.
After this tiny change the program worked.

However, I faced some issues:

### Error messages on stderr at startup

When I captured stderr, I found that the OSGI runtime produces lots of error messages on startup in https://bugs.eclipse.org/bugs/show_bug.cgi?id=547526. In my case they are looking like this:

org.eclipse.equinox.internal.p2.extensionlocation.SiteListener@666f4066: java.lang.UnsupportedOperationException: Not able to create StateObjectFactory implementation: org.eclipse.osgi.internal.resolver.StateObjectFactoryImpl.

I have no idea what these messages mean and why there are so many of them.

### Footprint

The disk footprint is significantly bigger (+ ~30 MB). Not really a problem for me. Maybe the memory footprint is also bigger, didn't investigate that...

### Files created at runtime

The OSGI runtime creates log files and lots of other files in ReportEngine\platform\configuration
and who knows where else?
Those files are not deleted at shutdown. New files are added with each start.
This is problematic for two different reasons:

1) This "disk memory leak" will cause to disk to run out of space sooner or later (probably much later). This is a no-go for an application that runs on an application server machine. 

2) If several instances of the BIRT runtime program run (as is the case in our application), the fact that those processes are creating files in the same directory is problematic (from a high-level point-of-view).

By setting the java.io.tmpdir system property to a fresh directory I can direct the temporary files that the BIRT runtime itself creates (e.g. for cached data from data sets) to that directory and clean up that directory afterwards, no problem.

However, whatever I tried (with other system properties like eclipse.p2.data.area, osgi.configuration.area, osgi.instance.area, org.osgi.framework.storage), the OSGI runtime will always create files in ReportEngine\platform\configuration.

By setting this directory read-only (recursively), I can avoid this - but I don't know if that has any negative consequences

**So to summarize, one needs to add some house-keeping code to the program and it is not clear if it is safe to run several BIRT OSGI processes in parellel.**

### Class-Path issues in own code

In our own code, we are using the iText library provided by BIRT for adding page numbering after BIRT has created the PDF.
This fails now with java.lang.ClassNotFoundException: org.eclipse.birt.report.engine.layout.pdf.font.FontMappingConfig:

java.lang.NoClassDefFoundError: org/eclipse/birt/report/engine/layout/pdf/font/FontMappingConfig
            at java.base/java.lang.ClassLoader.defineClass1(Native Method)
            at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
            at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
            at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:800)
            at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:698)
            at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:621)
            at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:579)
            at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
            at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
            at com.tp.server.lisa.birt.report.BirtEngineManager.readFontsConfiguration(BirtEngineManager.java:143)
            at com.tp.server.lisa.birt.report.BirtEngineManager.init(BirtEngineManager.java:68)
            at com.tp.server.lisa.birt.report.ServerMain.startEngineManager(ServerMain.java:365)
            at com.tp.server.lisa.birt.report.ServerMain.listen(ServerMain.java:208)
            at com.tp.server.lisa.birt.report.ServerMain.main(ServerMain.java:81)
Caused by: java.lang.ClassNotFoundException: org.eclipse.birt.report.engine.layout.pdf.font.FontMappingConfig
            at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
            at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
            at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
            ... 14 more

I think this is because iText was available in ReportEngine\lib with the POJO runtime as a normal JAR, but with the OSGI runtime, it is available as an OSGI jar in ReportEngine\platform\plugins. So maybe I can remedy this by adding that OSGI jar to the class path explicitly or by using a custom class loader or whatever.

### Questions

Does anyone know how to redirect the temporary files (`.manager` directories with `*.instance` files) to a different folder than ReportEngine\platform\configuration by specifying environment variables, system properties or the like?

Are the negative consequences of setting ReportEngine\platform\configuration read-only?

Or is it safe to run several OSGI processes using the same directory ReportEngine\platform\configuration?

Does the framework provide house-keeping utilities for that folder?

Do the error messages at stderr on startup mean that something will not work? Can they be avoided?

Is there a recommended way to access classes from the ReportEngine\platform directory (like iText) in one's own code (which also integrates BIRT OSGI)? Maybe it is possible to access iText classes using some BIRT/Eclipse/OSGI API between Platform.startup() and Platform.shutdown()?

**Comment by wimjongman** [comment]
## Redirect
You have to pass the correct osgi parameters [1]. Take a look at the report engine ReportRunner class to see how osgi is started.

[1] https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html

## Are the negative consequences of setting ReportEngine\platform\configuration read-only?
The osgi can't save state

## Houskeeping
Normally not needed

## Is there a recommended way to access classes 
Yes you have to add your bundle to plugins. You could even try to create a dropins folder in the platform directory. Maybe you have to add your bundle to the startup config in the config.ini

You have to dive into OSGi. This is not just a simple Q&A

**Comment by hvbtup** [comment]
Thanks for your answers, Wim.

> You have to pass the correct osgi parameters

Well, that's what I tried. 
Your hint looking at ReportRunner.java was helpful in a way.
The only difference between our initialization code and that used in ReportRunner.java (inside the method createEngineConfig()) is that we don't use setTempDir. But we set the java.io.tmpdir, so the effect should be the same (see PatformConfig.getTempDir()).

However, looking at PlatformConfig.java, I found the following comment:

	/**
	 * set the configuration used by the OSGi framework. The configuration includes
	 * all valid osgi configs except some reserved as follow:
	 * <li>1. osgi.install.area, using the one defined by BIRT_HOME.</li>
	 * <li>2. eclipse.ignoreApp, be true.</li>
	 * <li>3. osgi.noShutDown, be true.</li>
	 * <li>4. osgi.framework, using the one defined by BIRT_HOME</li>
	 * <li>5. osgi.framework.useSystemProperties, be false.</li>
	 * 
	 * @param osgiConfigMap
	 */
	public void setOSGiConfig(Map osgiConfigMap) {
		properties.put(OSGI_CONFIGURATION, osgiConfigMap);
	}

The comment could probably explain why specifying other locations per SystemProperties didn't have any effect.
But who can explain the comment itself - why it is intended that way and where is that implemented? 
I tried to understand it and found this very old commit https://github.com/eclipse/birt/commit/269e144660eeebb8b1e9f2a003d830351a3043fd which seems involved.

In particular org\eclipse\birt\core\framework\osgi\OSGILauncher.java contains a lot of logic which I fail to understand, nor do I understand the reasons for this. Anyway, this logic seems to force the temporary files to be created in ReportEngine\platform\configuration.

**Comment by wimjongman** [comment]
I think the idea is to start that OSGi runtime once. After it is started it will keep running in that process which makes the rendering of addition reports faster. The pojo is meant to run every time in a new process.

**Reply by hvbtup** [reply]
> I think the idea is to start that OSGi runtime once. 

That would work without fiddling around with the OSGI settings, too, so this is probably not the reason.

> After it is started it will keep running in that process which makes the rendering of addition reports faster. 

This also applies to the POJO runtime.

> The pojo is meant to run every time in a new process.

No, that is definitely wrong.
Insofar the ReportRunner class and the genReport.bat/sh script is a bad example. It is clearly meant for testing purposes only, because of the significant startup overhead. 
People who want to use BIRT in real life should not not look at ReportRunner.java as a reference. It only shows the principle. Perhaps it would be best if we add a note to the help output which clearly states something like this:

    This script is for testing purposes only.
    It is slow because of the startup overhead.
    See https://eclipse.github.io/birt-website/docs/integrating/reapi for details.

And on that page, we should make this point more clear. The "API overview" is a bit misleading (e.g. it says "repeat steps 4-6 for the next report" but it is an unordered list (UL) not a orderer list, so is is not clear that the three points to repeat for each report are only:
- Open the report design
- Set the report parameters
- Call either RunAndRenderTask or RunTask, then RenderTask

This is crucial for performance.

Perhaps it should also be noted that RunAndRenderTask is faster and uses less disk space than separate RunTask and RenderTask.

**Comment by TheLaserNinja** [comment]
A couple of answers.

1. The temporary files are not removed for a reason. In certain commercial applications, there is a retention period. This was configurable in the iHub and the iHub itself was responsible for the removal of any temporary files. So in the open source Eclipse BIRT, this is left to the application implementor.
2. Don't run multiple OSGi instances on the same directory. And don't run multiple engines. This will lead to issues. The Report Engine itself is already multi-threaded. It was designed to create multiple Run tasks and Render tasks as separate threads depending on the threading model used. 
3. Classes loaded in the OSGi framework were designed to operate in the OSGi framework, and trying to access them in your own application will lead to errors. They are completely different class-loaders. There was some black magic with the way the engine was designed to even be executed from a POJO application when using the OSGi runtime. I never bothered to look into it, I just considered it black-box and left it alone. This is also part of why the POJO runtime was created.

**Comment by hvbtup** [comment, marked as answer]
@TheLaserNinja Thank you for your answers.

Regarding the temporary files, note that I distinguish between those files that the data tools use to cache data (can be configured with java.io.tmpdir) and those that are OSGI specific (under platform).

Summarizing your answers and those from Wim, it seems that the OSGI runtime is quite different from the POJO runtime.
Since the POJO runtime is working as of today, I won't need the OSGI runtime and will continue to use the POJO runtime.

I'll try to summarize the differences here:

## Differences between OSGI runtime and PJO runtime

| Topic                                            | OSGI runtime                  | POJO runtime |
|-----------------------------------|---------------------------|-----------------|
| Integration into an application     | application must be OSGI-based | "classic" application  |
| Multiple proceccess possible        | not supported                  | supported if each process uses a separate java.io.tmpdir |
| Supports multi-threadeding per process         | yes                                   | yes |
| House-keeping responsibility       | app (java.io.tmpdir + platform/configuration) | app (java.io.tmpdir) |

The application (or a wrapper program for it) should delete the temporary files (which BIRT created) at a suitable time, for example before the OSGI startup or after the OSGI shutdown or before the application is started or after it is closed.

## Why use multiple processes at all, you might ask?

A valid reason for using multiple BIRT processes instead of just multiple BIRT threads is to avoid issues with otherwise unrelated reports when one report causes VM problems. In particular, OOM.
Consider the following scenario in a single-process BIRT server implementation which I saw a few times in production environments.

You start some small reports (eg for labels) and one or two memory-intensive reports (e.g. with a lot of images) in parallel.
Memory exhaustion will first slow down everything significantly because the VM will often have to run a GC to make a few more bytes available. 
Then one or more of the reports (the big reports **or the small reports**) will run into OOM when they ask the JVM for a few bytes more, just because the big report(s) used nearly all the JVM memory.
This is very hard to explain to users.

That's why we switched from a single-process-multi-threaded BIRT server to a multi-process-single-threaded server.
An OOM in a big report will thus not effect small reports.

Of course this has some disadvantages as well, for example the total RAM used is higher than with a single process, which can lead to out-of-memory on the OS (not Java) level.

A better solution would be a properly configured multi-process-multi-threaded server (where the max. number of threads per process depends on the kind of report, e.g. a one process can easily create a dozen or so label reports).

**Comment by TheLaserNinja** [comment]
One thing to clarify.The application doesn't need to be OSGi based for integration. It just needs to call Platform.startup() so the engines internal OSGi plug-ins will start. This is why you only ever call Platform.startup() once. But the OSGi runtime can be integrated into a POJO Java or web servlet application. In fact, the commercial iHub uses the OSGi runtime in favor or the POJO runtime. You can run multiple platforms, but you might get some interesting results. The best practice was to only ever run a single platform and a single engine per server. But thats not to say youre wrong about having seperate instances running. The single instance was more a rule of thumb. This gets into discussions around synchronous and asynchronous execution, dedicated engine for off-hour large report execution, factories, etc. To meet certain SLAs for long report execution dedicated instances or factories would be stood up to run reports at a scheduled time that would take upwards of 8 hours to a weekend to run, separate from engine instances that serviced on-demand reports. But now the scope of the discussion is getting into some deep territory in terms of sizing and threading. The short answer is increase the memory size. Used to be the biggest out of memory issue was with MaxPermSize, but I think that was done away with (and isnt an issue with the POJO runtime). But there are sizing considerations to take into account, such as RAM, CPU, etc. But if a report is causing OOM errors, my first stop would be to optimize the report. This isn't always possible. I used to have a tool that would benchmark reports so you could track down where a report was getting hung up. Let you select different threading models so you could see how it changed overall performance.

-John

**Comment by Markus-Elfring** [comment]
>`org.eclipse.equinox.internal.p2.extensionlocation.SiteListener@666f4066: java.lang.UnsupportedOperationException: Not able to create StateObjectFactory implementation: org.eclipse.osgi.internal.resolver.StateObjectFactoryImpl`

:crystal_ball:
* Will development interests grow any more for affected implementation details?
* How will the chances evolve for the support of file generation by such programming interfaces?

**Reply by wimjongman** [reply]
https://github.com/eclipse/birt/issues/835
